pkg/kubelet/stats/helper.go

```diff
@@ -270,6 +270,31 @@ func latestContainerStats(info *cadvisorapiv2.ContainerInfo) (*cadvisorapiv2.Con
 	return latest, true
 }
 
+// adjustCgroupv2NodeMemoryUsage adjusts memory usage on v2 for cadvisor.
+// runc returns `total - free` from /proc/meminfo as usage to cadvisor for v2.
+// This produces a substantially higher value for root cgroup memory usage
+// for v2 compared to v1 on machines with comparable memory usage.
+// Adjust the calculation by manually summing `rss` and `cache` to align
+// v1 + v2.
+func adjustCgroupv2NodeMemoryUsage(info *cadvisorapiv2.ContainerInfo) {
+	for idx := range info.Stats {
+		stat := info.Stats[idx]
+		oldUsage := stat.Memory.Usage
+		newUsage := stat.Memory.RSS + stat.Memory.Cache
+		diff := oldUsage - newUsage // amount to adjust working set
+		stat.Memory.Usage = newUsage
+		stat.Memory.WorkingSet -= diff
+		// cadvisor calculates swap as:
+		// `ret.Memory.Swap = s.MemoryStats.SwapUsage.Usage - s.MemoryStats.Usage.Usage`
+		// so if we adjust usage by `diff`, we need to adjust swap by `diff` as well.
+		// from cadvisor, swap = swap usage - mem usage
+		// for v2 this is (total - free swap) - (total - free mem)
+		// with 0 total and 0 free swap this reduces to swap = - (total - free mem)
+		// do it anyway for consistency.
+		stat.Memory.Swap += diff
```

Comment:
I'm not sure this is desirable. It will lead to kubelet reporting swap usage on machines with no swap.

it's consistent with runc pod cgroup calculations. but v2 usage from runc does not include swap. runc adds it back for non-root cgroups. https://github.com/opencontainers/runc/blob/c99e6c6c5d6dd27158993f57ab767f618acc35f2/libcontainer/cgroups/fs2/memory.go#L119-L122

for root cgroup, cadvisor does _not_ add usage back to swap -- which I feel is another inconsistency. but I'm tackling one problem at a time...
https://github.com/opencontainers/runc/blob/c99e6c6c5d6dd27158993f57ab767f618acc35f2/libcontainer/cgroups/fs2/memory.go#L108-L110
https://github.com/opencontainers/runc/blob/c99e6c6c5d6dd27158993f57ab767f618acc35f2/libcontainer/cgroups/fs2/memory.go#L214-L215

Reply:
I dropped the `stat.Memory.Swap += diff` for now and left a longer comment.